src/Interactive/Features/Interactive/Core/InteractiveHost.RemoteService.cs

```diff
@@ -113,38 +118,47 @@ private void ReadOutput(bool error)
 
             internal void Dispose(bool joinThreads)
             {
-                // set _disposing so that we don't attempt restart the host anymore:
-                _disposing = true;
+                // There can be a call from host initiated from OnProcessExit. 
+                // This check on the beginning helps to avoid the circular reference.
+                if (_processExitHandling == ProcessExitHooked)
+                {
+                    using (_disposeSemaphore.DisposableWait())
+                    {
+                        if (_processExitHandling == ProcessExitHooked)
+                        {
+                            Process.Exited -= ProcessExitedHandler;
+                            _processExitHandling = ProcessExitHandled;
+                        }
+                    }
+                }
 
                 InitiateTermination(Process, _processId);
 
-                // only tests require joining the threads, so we can wait synchronously
-                if (joinThreads)
+                var readOutputThreadJoinTask = Task.Run(() =>
```

Comment:
>var readOutputThreadJoinTask = Task.Run(() => [](start = 16, length = 45)

Why do we need to do this in a task?

Reply:
Discussed offline. Agreed to keep is sync until we'll see any issues with this.

---
In reply to: [163378489](https://github.com/dotnet/roslyn/pull/24058#discussion_r163378489) [](ancestors = 163378489)